### PR TITLE
Update CONTRIBUTING.rst with docstring guidelines

### PR DESCRIPTION
**Pull request by akaszynski**
Address [comments](https://github.com/pyvista/pyvista/pull/3457#issuecomment-1278021788) made by @adeak:

> > I'll also keep this in mind when adding any new features or updating docs (as well as remind others to do so).
> 
> Documenting this internal convention in our contributing guide would go a long way in replacing the mixed style in the long run, @akaszynski.

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
-  `numpydoc Style Guide`_.
+PyVista follows the ``numpydoc`` style for its docstrings. Please follow the
+`numpydoc Style Guide`_ in all ways except for the following:
+
 * Be sure to describe all ``Parameters`` and ``Returns`` for all public
   methods.
 * We strongly encourage you to add an example section. PyVista is a visual
   library, so adding examples that show a plot will really help users figure
   out what individual methods do.
+* With optional parameters, use ``default: <value>`` instead of ``optional``
```
Comment (adeak) [inline CONTRIBUTING.rst:217]: In https://github.com/pyvista/pyvista/pull/3457/files we seemed to use `default=<value>`. Either works for me, we should just make sure we choose one on purpose.

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+  when the parameter has a default value instead of ``None``.
+
+Sample docstring follows:
```
Comment (adeak) [inline CONTRIBUTING.rst:220]: Good call.

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+
+.. code:: python
+
+    def slice_x(self, x=None, generate_triangles=False):
+        """Create an orthogonal slice through the dataset in the x-direction.
```
Comment (adeak) [inline CONTRIBUTING.rst:225]: Nitpick, but this example should be exemplary.
```suggestion
        """Create an orthogonal slice through the dataset in the X direction.
```

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+
+        Yields a MutliBlock dataset of the three slices.
```
Comment (adeak) [inline CONTRIBUTING.rst:227]:
```suggestion
        Yields a ``MultiBlock`` dataset of the three slices.
```
Or maybe even showcase intersphinx?
```suggestion
        Yields a :class:`MultiBlock <pyvista.MultiBlock>` dataset of the three slices.
```
(assuming this is within the usual length.)

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+
+        Parameters
+        ----------
+        x : float, optional
+            The X location of the YZ slice.
```
Comment (adeak) [inline CONTRIBUTING.rst:232]: For optional params we should ideally always explain the default behaviour.

File: CONTRIBUTING.rst [hunk 6]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+
+        generate_triangles : bool, default: False
+            If this is enabled, the output will be triangles. Otherwise the output
+            will be the intersection polygons.
```
Comment (adeak) [inline CONTRIBUTING.rst:236]:
```suggestion
            If this is enabled, the output will be all triangles. Otherwise the output
            will consist of the intersection polygons.
```

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+
+        Returns
+        -------
+        pyvista.PolyData
```
Comment (adeak) [inline CONTRIBUTING.rst:240]: `PolyData` or `MultiBlock`?

File: CONTRIBUTING.rst [hunk 8]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+        See :ref:`slice_example` for more examples using this filter.
+
+        """
+
+Note the following:
+
+* The parameter definition of ``generate_triangles`` uses ``default: False``,
+  and does not include the default in the docstring's "description" section.
+* There is a newline between each parameter. This is different than
+  ``numpydoc``'s documentation where there are no lines between parameter
```
Comment (adeak) [inline CONTRIBUTING.rst:261]:
```suggestion
  ``numpydoc``'s documentation where there are no empty lines between parameter
```

File: CONTRIBUTING.rst [hunk 9]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+  doc strings.
+* This docstring contains both a returns section and examples section.
```
Comment (adeak) [inline CONTRIBUTING.rst:263]:
```suggestion
  docstrings.
* This docstring also contains a returns section and an examples section.
```

File: CONTRIBUTING.rst [hunk 10]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+* Returns section does not include the parameter name if it has a single return.
+  Multiple returns (not shown) should have descriptive parameter names.
+* The examples section references the "full example" if it exists.
```
Comment (adeak) [inline CONTRIBUTING.rst:266]: Not sure this is the best possible phrasing, feel free to adjust.
```suggestion
* The returns section does not include the parameter name if the function has 
  a single return value. Multiple return values (not shown) should have 
  descriptive parameter names for each returned value, in the same format as
  the input parameters.
* The examples section references the "full example" in the gallery if it 
  exists.
```

**Review by adeak — COMMENTED** [review]
Thanks for adding this. I've picked many nits because this is something that we expect users to follow, so we should be as precise as possible.

File: CONTRIBUTING.rst (hunk 4 above)
Comment (MatthewFlamm) [reply CONTRIBUTING.rst:227]: I think this whole line can be deleted.  This docstring stub is presumably referencing a slice operation in the x-normal direction, i.e. it only yields a single PolyData slice and could be described more in the `Returns` section if needed.

File: CONTRIBUTING.rst [hunk 11]
```diff
@@ -203,16 +203,67 @@ Docstrings
...
+        pyvista.PolyData
+            Sliced dataset.
+
+        Examples
+        --------
+        Slice the random hills dataset with one orthogonal plane.
+
+        >>> from pyvista import examples
+        >>> hills = examples.load_random_hills()
+        >>> slices = hills.slice_orthogonal(5, generate_triangles=False)
```
Comment (MatthewFlamm) [inline CONTRIBUTING.rst:249]:
```suggestion
        >>> slices = hills.slice_x(5, generate_triangles=False)
```

File: CONTRIBUTING.rst (hunk 1 above)
Comment (MatthewFlamm) [reply CONTRIBUTING.rst:217]: The current code base has more `default: <value>`.  I found 164 hits with this. 
 This includes some false positives, but not as many as I would have suspected.  I found 36 hits for `default=<value>`, with also some false positives.  Before #3457, it would have leaned even heavier towards the colon usage.  Although, I'm fine either way as well.

Comment (adeak) [reply CONTRIBUTING.rst:217]: I'd also prefer a colon here, ignoring all legacy considerations.

Comment (akaszynski) [reply CONTRIBUTING.rst:217]: Agree with the colon. Let's go with that and consider a refactor. At least it's low hanging.

File: CONTRIBUTING.rst (hunk 7 above)
Comment (akaszynski) [reply CONTRIBUTING.rst:240]: This is just a fake function, so we can just go with PolyData.

File: CONTRIBUTING.rst [hunk 12]
```diff
@@ -203,16 +203,144 @@ Docstrings
...
+  descriptive parameter names for each returned value, in the same format as
+  the input parameters.
+* The examples section references the "full example" in the gallery if it 
+  exists.
+
+
+Depreciating Features or other Backwards-Breaking Changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When implementing backwards-breaking changes within PyVista, care must be taken
+to make the modifications non-breaking in one of the following ways:
```
Comment (akaszynski) [inline CONTRIBUTING.rst:274]:
```suggestion
When implementing backwards-breaking changes within PyVista, care must be taken
to give users the chance to adjust to any new changes. Any non-backwards
compatible modifications should proceed through the following steps:
```

File: CONTRIBUTING.rst [hunk 13]
```diff
@@ -203,16 +203,144 @@ Docstrings
...
+
+* Retain the old behavior and issue a ``PyVistaDeprecationWarning`` indicating
+  the new interface you should use.
+* Retain the old behavior but raise a ``pyvista.core.errors.DeprecationError``
+  indicating the new interface you must use.
+* Remove the old behavior.
```
Comment (akaszynski) [inline CONTRIBUTING.rst:280]:
```suggestion
# Retain the old behavior and issue a ``PyVistaDeprecationWarning`` indicating
  the new interface you should use.
# Retain the old behavior but raise a ``pyvista.core.errors.DeprecationError``
  indicating the new interface you must use.
# Remove the old behavior.
```

File: CONTRIBUTING.rst [hunk 14]
```diff
@@ -203,16 +203,144 @@ Docstrings
...
+
+Whenever possible, PyVista developers should seek to have at least three minor
+versions of backwards compatibility to give users the ability to update their
+software and scripts.
+
+Here's an example of soft depreciation. Note the usage of both
```
Comment (akaszynski) [inline CONTRIBUTING.rst:286]:
```suggestion
Here's an example of a soft depreciation of a function. Note the usage of both
```

File: CONTRIBUTING.rst [hunk 15]
```diff
@@ -203,16 +203,144 @@ Docstrings
...
+``PyVistaDeprecationWarning`` warning and the ``.. deprecated`` Sphinx
```
Comment (akaszynski) [inline CONTRIBUTING.rst:287]:
```suggestion
Here's an example of a soft depreciation of a function. Note the usage of both
the ``PyVistaDeprecationWarning`` warning and the ``.. deprecated`` Sphinx
```

File: CONTRIBUTING.rst [hunk 16]
```diff
@@ -203,16 +203,145 @@ Docstrings
...
+* This docstring also contains a returns section and an examples section.
+* The returns section does not include the parameter name if the function has 
+  a single return value. Multiple return values (not shown) should have 
+  descriptive parameter names for each returned value, in the same format as
+  the input parameters.
+* The examples section references the "full example" in the gallery if it 
+  exists.
+
+
+Depreciating Features or other Backwards-Breaking Changes
```
Comment (adeak) [inline CONTRIBUTING.rst:271]:
```suggestion
Deprecating Features or other Backwards-Breaking Changes
```

File: CONTRIBUTING.rst [hunk 17]
```diff
@@ -203,16 +203,145 @@ Docstrings
...
+   the new interface you should use.
+#. Retain the old behavior but raise a ``pyvista.core.errors.DeprecationError``
+   indicating the new interface you must use.
+#. Remove the old behavior.
+
+Whenever possible, PyVista developers should seek to have at least three minor
+versions of backwards compatibility to give users the ability to update their
+software and scripts.
+
+Here's an example of a soft depreciation of a function. Note the usage of both
```
Comment (adeak) [inline CONTRIBUTING.rst:287]:
```suggestion
Here's an example of a soft deprecation of a function. Note the usage of both
```

File: CONTRIBUTING.rst [hunk 18]
```diff
@@ -258,12 +257,91 @@ Note the following:
...
```
Comment (adeak) [inline CONTRIBUTING.rst:287]:
```suggestion
Here's an example of a soft deprecation of a function. Note the usage of both
```

File: CONTRIBUTING.rst [hunk 19]
```diff
@@ -258,12 +257,91 @@ Note the following:
...
+        b : float
+            Second term to add.
+
+        Returns
+        -------
+        float
+            Sum of the two inputs.
+
+        """
+        # depreciated 0.37.0, convert to error in 0.40.0, remove 0.41.0
```
Comment (adeak) [inline CONTRIBUTING.rst:313]:
```suggestion
        # deprecated 0.37.0, convert to error in 0.40.0, remove 0.41.0
```

File: CONTRIBUTING.rst [hunk 20]
```diff
@@ -258,12 +257,91 @@ Note the following:
...
+        PyVistaDeprecationWarning(
+            '`addition` has been deprecated. Use pyvista.add instead'
+        )
+        add(a, b)
+
+
+    def add(a, b):
+        """Add two numbers.
```
Comment (adeak) [inline CONTRIBUTING.rst:321]: I suspect the ellipsis was meant to be a placeholder for the docstring, but we need the ending triple quotes somewhere because this is breaking syntax highlighting at least on github. But if we put the """ after the ellipsis we'd probably have to add another ... or `pass` to get a syntactically valid example... might be simpler to have a docstring stub instead:
```suggestion
        """Add two numbers."""
```

File: pyvista/utilities/errors.py [hunk 21]
```diff
@@ -260,34 +259,73 @@ def __repr__(self):
...
-        text_width : int, optional
-            The text width for non-HTML display modes
+    sort : bool, default: False
+        Alphabetically sort the packages.
 
-        sort : bool, optional
-            Alphabetically sort the packages
+    gpu : bool, default: True
+        Gather information about the GPU. Defaults to ``True`` but if
+        experiencing renderinng issues, pass ``False`` to safely generate a
```
Comment (adeak) [inline pyvista/utilities/errors.py:281]:
```suggestion
        experiencing rendering issues, pass ``False`` to safely generate a
```

File: setup.py [hunk 22]
```diff
@@ -16,7 +16,6 @@
     'numpy',
     'imageio',
     'pillow',
-    'appdirs',
```
Comment (adeak) [inline setup.py:19]: Is this removal related?

File: CONTRIBUTING.rst [hunk 23]
```diff
@@ -203,16 +203,145 @@ Docstrings
...
+
+        ...
```
Comment (akaszynski) [inline CONTRIBUTING.rst:323]:
```suggestion
        pass  # implementation goes here
```

File: setup.py (hunk 22 above)
Comment (akaszynski) [reply setup.py:19]: This should have happened in https://github.com/pyvista/pyvista/pull/3218

File: CONTRIBUTING.rst [hunk 24]
```diff
@@ -268,7 +268,7 @@ Note the following:
   exists.
 
 
-Depreciating Features or other Backwards-Breaking Changes
+Deprecating Features or other Backwards-Breaking Changes
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (adeak) [inline CONTRIBUTING.rst:272]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: pyvista/utilities/helpers.py [hunk 25]
```diff
@@ -1306,15 +1306,15 @@ class ProgressMonitor:
...
+    message : str, default: ""
         Message to display in the progress bar.
 
     scaling : float, optional
         Unused keyword argument.
 
     """
 
-    def __init__(self, algorithm, message="", scaling=100):
+    def __init__(self, algorithm, message="", scaling=None):
```
Comment (adeak) [inline pyvista/utilities/helpers.py:1317]: If I'm reading things correctly, `scaling` isn't even used in this function. Why the change then, and can we just dump it? Or is this here for backwards compat?

File: pyvista/utilities/regression.py [hunk 26]
```diff
@@ -64,12 +64,12 @@ def compare_images(im1, im2, threshold=1, use_vtk=True):
...
         window, or ``vtkImageData``.
 
-    threshold : int, optional
+    threshold : int, default: 1
         Threshold tolerance for pixel differences.  This should be
         greater than 0, otherwise it will always return an error, even
         on identical images.
 
-    use_vtk : bool, optional
+    use_vtk : bool, bool: True
```
Comment (adeak) [inline pyvista/utilities/regression.py:72]:
```suggestion
    use_vtk : bool, default: True
```

File: pyvista/utilities/helpers.py (hunk 25 above)
Comment (adeak) [reply pyvista/utilities/helpers.py:1317]: Ah, it's there in the docstring. Sorry. I understand then. Perhaps we should remove it eventually...

Comment (akaszynski) [reply pyvista/utilities/helpers.py:1317]: It's correct that it's not used. I'm assuming that it's been kept here for backwards compatibility, so probably the right thing to do is to warn then raise during a later release.

It's a bit out of the scope of this PR and I should have left it alone.

File: CONTRIBUTING.rst [hunk 27]
```diff
@@ -203,16 +203,145 @@ Docstrings
...
+        Slice the random hills dataset with one orthogonal plane.
+
+        >>> from pyvista import examples
+        >>> hills = examples.load_random_hills()
+        >>> slices = hills.slice_x(5, generate_triangles=False)
+        >>> slices.plot(line_width=5)
+
+        See :ref:`slice_example` for more examples using this filter.
+
+        """
```
Comment (adeak) [inline CONTRIBUTING.rst:253]:
```suggestion
        """

        pass  # implementation goes here
```

**Review by adeak — APPROVED** [review]
LGTM now, thanks!
